### PR TITLE
add --seed option to --createRandomTest

### PR DESCRIPTION
**Pull request by winsvega**
usage

```
./testeth -t GeneralStateTests -- --createRandomTest --seed 123457
```

### CODE REVIEW

File: test/tools/fuzzTesting/fuzzHelper.h [hunk 1]
```diff
@@ -47,14 +48,20 @@ class RandomCode
...
 	RandomCode(RandomCode const&) = delete;
 	void operator=(RandomCode const&) = delete;
+	~RandomCode() { delete generator; }
 
 private:
-	RandomCode(){}
-	BoostRandomCode generator;
+	RandomCode(){
+		if (Options::get().randomTestSeed.empty())
+			generator = new BoostRandomCode();
```
Comment (chfast) [inline test/tools/fuzzTesting/fuzzHelper.h]: This abstraction is stupid. Only use DeterministicRandomCode (with better name).
If seed is empty just pick a random one (you can use random_device for it).

File: test/tools/fuzzTesting/fuzzHelper.h [hunk 2]
```diff
@@ -47,14 +48,20 @@ class RandomCode
 	static RandomCodeBase& get()
 	{
 		static RandomCode instance;
-		return instance.generator;
+		return *instance.generator;
 	}
 	RandomCode(RandomCode const&) = delete;
 	void operator=(RandomCode const&) = delete;
+	~RandomCode() { delete generator; }
```
Comment (chfast) [inline test/tools/fuzzTesting/fuzzHelper.h]: No kidding. This is no go. Use `std::unique_ptr`. But probably not needed at all according to my other comments.

File: test/tools/libtesteth/Options.cpp [hunk 3]
```diff
@@ -205,6 +206,11 @@ Options::Options(int argc, char** argv)
 		}
 		else if (arg == "--createRandomTest")
 			createRandomTest = true;
+		else if (arg == "--seed")
+		{
+			throwIfNoArgumentFollows();
+			randomTestSeed = argv[++i];
```
Comment (chfast) [inline test/tools/libtesteth/Options.cpp]: I think the seed should be a number, not a string. Preferably `uint64_t` as [mt19937_64](http://en.cppreference.com/w/cpp/numeric/random/mersenne_twister_engine) is good choice for random engine.

File: test/tools/fuzzTesting/fuzzHelper.h (hunk 1 above)
Comment (winsvega) [reply test/tools/fuzzTesting/fuzzHelper.h]: this is just an example. there will be a different implementation of DeterministicRandomCode

File: test/tools/libtesteth/Options.h [hunk 4]
```diff
@@ -54,6 +54,7 @@ class Options
 	bool statediff = false;///< Fill full post state in General tests
 	bool fulloutput = false;///< Replace large output to just it's length
 	bool createRandomTest = false; ///< Generate random test
+	uint64_t randomTestSeed; ///< Define a seed for random test
```
Comment (chfast) [inline test/tools/libtesteth/Options.h]: Init it.

File: test/tools/libtesteth/Options.cpp [hunk 5]
```diff
@@ -205,6 +207,11 @@ Options::Options(int argc, char** argv)
 		}
 		else if (arg == "--createRandomTest")
 			createRandomTest = true;
+		else if (arg == "--seed")
+		{
+			throwIfNoArgumentFollows();
+			randomTestSeed = static_cast<uint64_t>(toInt(argv[++i]));
```
Comment (chfast) [inline test/tools/libtesteth/Options.cpp]: I think you should fails also if the value is bigger than uint64_t.

File: test/tools/fuzzTesting/BoostRandomCode.cpp [hunk 6]
```diff
@@ -38,9 +38,14 @@ u256 BoostRandomCode::randomUniInt(u256 const& _minVal, u256 const& _maxVal)
 
 void BoostRandomCode::refreshSeed()
 {
-	auto now = std::chrono::steady_clock::now().time_since_epoch();
-	auto timeSinceEpoch = std::chrono::duration_cast<std::chrono::nanoseconds>(now).count();
-	gen.seed(static_cast<unsigned int>(timeSinceEpoch));
+	if (Options::get().randomTestSeed == 0)
+	{
+		auto now = std::chrono::steady_clock::now().time_since_epoch();
```
Comment (chfast) [inline test/tools/fuzzTesting/BoostRandomCode.cpp]: Why have you used a clock instead of `std::random_device`?

File: test/tools/fuzzTesting/BoostRandomCode.cpp [hunk 7]
```diff
@@ -38,9 +38,14 @@ u256 BoostRandomCode::randomUniInt(u256 const& _minVal, u256 const& _maxVal)
...
+		auto timeSinceEpoch = std::chrono::duration_cast<std::chrono::nanoseconds>(now).count();
+		gen.seed(static_cast<unsigned int>(timeSinceEpoch));
```
Comment (chfast) [inline test/tools/fuzzTesting/BoostRandomCode.cpp]: This should be cast to `gen::result_type`.

File: test/tools/fuzzTesting/fuzzHelper.h [hunk 8]
```diff
@@ -47,14 +47,16 @@ class RandomCode
...
+		return *instance.generator.get();
 	}
 	RandomCode(RandomCode const&) = delete;
 	void operator=(RandomCode const&) = delete;
 
 private:
-	RandomCode(){}
-	BoostRandomCode generator;
+	RandomCode(){
+		generator.reset(new BoostRandomCode());
```
Comment (chfast) [inline test/tools/fuzzTesting/fuzzHelper.h]: I don't see why this changes is needed now. Can we keep the old composition?

File: test/tools/fuzzTesting/fuzzHelper.h [hunk 9]
```diff
@@ -47,14 +47,16 @@ class RandomCode
 	static RandomCodeBase& get()
 	{
 		static RandomCode instance;
-		return instance.generator;
+		return *instance.generator.get();
```
Comment (chfast) [inline test/tools/fuzzTesting/fuzzHelper.h]: You don't need `.get()` here.

File: test/tools/libtesteth/Options.h (hunk 4 above)
Comment (pirapira) [inline test/tools/libtesteth/Options.h]: Can you turn this into `boost::optional<uint64_t>`?  When this option is not given it shouldn't be zero, or any number.

File: test/tools/fuzzTesting/fuzzHelper.h [hunk 10]
```diff
@@ -47,14 +47,16 @@ class RandomCode
...
 	}
 	RandomCode(RandomCode const&) = delete;
 	void operator=(RandomCode const&) = delete;
 
 private:
-	RandomCode(){}
-	BoostRandomCode generator;
+	RandomCode(){
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.h]: The `{` should be on the next line.

File: test/tools/fuzzTesting/fuzzHelper.h [hunk 11]
```diff
@@ -47,14 +47,16 @@ class RandomCode
...
+		generator.reset(new BoostRandomCode());
+	}
+	std::unique_ptr<RandomCodeBase> generator;
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.h]: Why not just have
```
RandomCodeBase generator;
```
?

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 12]
```diff
@@ -257,32 +257,36 @@ std::string RandomCodeBase::generate(int _maxOpNumber, RandomCodeOptions const&
...
 		eth::InstructionInfo info = eth::instructionInfo(inst);
 
-		if (info.name.find("INVALID_INSTRUCTION") != std::string::npos || info.name.empty()
+		while (info.name.find("INVALID_INSTRUCTION") != std::string::npos || info.name.empty()
 			|| std::find(invalidOpcodes.begin(), invalidOpcodes.end(), inst) != invalidOpcodes.end())
 		{
 			if (_options.useUndefinedOpCodes)
+			{
 				code += toCompactHex(opcode, 1);
+				return "0x" + code;
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: Why not `continue` and fill more opcodes?

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 13]
```diff
@@ -257,32 +257,36 @@ std::string RandomCodeBase::generate(int _maxOpNumber, RandomCodeOptions const&
...
+			}
 			else
 			{
 				//Byte code is yet not implemented. do not count it.
-				i--;
+				opcode++;
+				inst = (eth::Instruction) opcode;
+				info = eth::instructionInfo(inst);
 				continue;
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: Why is `code` changed in all other cases but not here?

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 14]
```diff
@@ -257,32 +257,36 @@ std::string RandomCodeBase::generate(int _maxOpNumber, RandomCodeOptions const&
 		eth::Instruction inst = (eth::Instruction) opcode;
 		eth::InstructionInfo info = eth::instructionInfo(inst);
 
-		if (info.name.find("INVALID_INSTRUCTION") != std::string::npos || info.name.empty()
+		while (info.name.find("INVALID_INSTRUCTION") != std::string::npos || info.name.empty()
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: Why is this a loop now, what is it iterating over?  What changes between the first run and the second run?

File: test/tools/fuzzTesting/fuzzHelper.h (hunk 11 above)
Comment (winsvega) [reply test/tools/fuzzTesting/fuzzHelper.h]: I plan to add one more implementation

File: test/tools/fuzzTesting/fuzzHelper.cpp (hunk 14 above)
Comment (winsvega) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: random codes are genereated in for sycle. 
if a random code is undefined  then we change it to a defined opcode in a while loop untill it is defined. (if we dont accept undefined opcodes)

there was an issue with static seed. a random undefined opcode would lead to endless loop.

File: test/tools/fuzzTesting/fuzzHelper.cpp (hunk 12 above)
Comment (winsvega) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: it is continued in a for loop

File: test/tools/fuzzTesting/fuzzHelper.cpp (hunk 13 above)
Comment (winsvega) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: random codes are genereated in for sycle.
if a random code is undefined then we change it to a defined opcode in a while loop untill it is defined. (if we dont accept undefined opcodes)

there was an issue with static seed. a random undefined opcode would lead to endless loop.

File: test/tools/fuzzTesting/BoostRandomCode.cpp (hunk 7 above)
Comment (winsvega) [reply test/tools/fuzzTesting/BoostRandomCode.cpp]: this page says it has decreasing perfomance 
http://en.cppreference.com/w/cpp/numeric/random/random_device

File: test/tools/fuzzTesting/BoostRandomCode.cpp (hunk 6 above)
Comment (winsvega) [reply test/tools/fuzzTesting/BoostRandomCode.cpp]: this page says it has decreasing perfomance 
http://en.cppreference.com/w/cpp/numeric/random/random_device

File: test/tools/fuzzTesting/BoostRandomCode.cpp (hunk 7 above)
Comment (winsvega) [reply test/tools/fuzzTesting/BoostRandomCode.cpp]: there is no such thing

File: test/tools/fuzzTesting/fuzzHelper.cpp (hunk 12 above)
Comment (pirapira) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: No, when you say `return`, the execution returns from the function.

You should continue in the for-loop.  Do you know `break` in loops?

File: test/tools/fuzzTesting/fuzzHelper.cpp (hunk 13 above)
Comment (pirapira) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: I suggest separating this `findingDefinedOpcode` feature as a function.

File: test/tools/fuzzTesting/fuzzHelper.h [hunk 15]
```diff
@@ -47,14 +47,17 @@ class RandomCode
...
 	void operator=(RandomCode const&) = delete;
 
 private:
-	RandomCode(){}
-	BoostRandomCode generator;
+	RandomCode()
+	{
+		generator.reset(new BoostRandomCode());
+	}
+	std::unique_ptr<RandomCodeBase> generator;
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.h]: Why is this not `RandomCodeBase generator;`?

Comment (winsvega) [reply test/tools/fuzzTesting/fuzzHelper.h]: Because there will be another implementation other than BoostRandomCode

File: test/tools/fuzzTesting/fuzzHelper.cpp (hunk 12 above)
Comment (winsvega) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: hm your right. but it should exit the while loop and continue the for loop then

File: test/tools/fuzzTesting/fuzzHelper.h (hunk 15 above)
Comment (pirapira) [reply test/tools/fuzzTesting/fuzzHelper.h]: OK.

File: test/tools/fuzzTesting/fuzzHelper.cpp (hunk 12 above)
Comment (pirapira) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: Check what `break` does in loops.

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 16]
```diff
@@ -239,52 +239,58 @@ std::string RandomCodeBase::rndByteSequence(int _length, SizeStrictness _sizeTyp
...
-		eth::Instruction inst = (eth::Instruction) opcode;
-		eth::InstructionInfo info = eth::instructionInfo(inst);
-
-		while (info.name.find("INVALID_INSTRUCTION") != std::string::npos || info.name.empty()
-			|| std::find(invalidOpcodes.begin(), invalidOpcodes.end(), inst) != invalidOpcodes.end())
+		if (!isOpcodeDefined(opcode) && _options.useUndefinedOpCodes)
 		{
-			if (_options.useUndefinedOpCodes)
+			code += toCompactHex(opcode, 1);
+			continue;
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: Here, the execution enters an infinite loop.

I think the execution should go out of the while loop.

Comment (winsvega) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: this continue goes to the for loop. it wont be infinite

File: test/tools/fuzzTesting/fuzzHelper.h (hunk 15 above)
Comment (chfast) [reply test/tools/fuzzTesting/fuzzHelper.h]: I still don't agree on this. Introduce this change when introducing the other implementation of RandomCodeBase.

Comment (winsvega) [reply test/tools/fuzzTesting/fuzzHelper.h]: there was. during the refactoring of this PR we decided to remove it.

File: test/tools/fuzzTesting/fuzzHelper.cpp (hunk 16 above)
Comment (pirapira) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: I think you are mistaken what `continue` does in nested loops.

Comment (winsvega) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: I have removed the nested loop. there is makeOpcodeDefined func now

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 17]
```diff
@@ -239,52 +239,58 @@ std::string RandomCodeBase::rndByteSequence(int _length, SizeStrictness _sizeTyp
 	return hash;
 }
 
+bool isOpcodeDefined(uint8_t _opcode)
+{
+	eth::Instruction inst = (eth::Instruction) _opcode;
+	eth::InstructionInfo info = eth::instructionInfo(inst);
+	if (info.name.find("INVALID_INSTRUCTION") != std::string::npos || info.name.empty()
```
Comment (chfast) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: `if (...) return true; else return false;` is not very smart. Just return check condition value.

File: test/tools/libtesteth/Options.cpp [hunk 18]
```diff
@@ -210,7 +209,7 @@ Options::Options(int argc, char** argv)
 		else if (arg == "--seed")
 		{
 			throwIfNoArgumentFollows();
-			randomTestSeed = static_cast<uint64_t>(toInt(argv[++i]));
+			randomTestSeed = static_cast<uint64_t>(min<u256>(UINT64_MAX, toInt(argv[++i])));
```
Comment (chfast) [inline test/tools/libtesteth/Options.cpp]: What's the difference? You sill hide the invalid value error.

File: test/tools/libtesteth/Options.cpp [hunk 19]
```diff
@@ -278,7 +277,7 @@ Options::Options(int argc, char** argv)
 	}
 	else
 	{
-		if (randomTestSeed != 0)
+		if (randomTestSeed.is_initialized())
```
Comment (chfast) [inline test/tools/libtesteth/Options.cpp]: It can be `if (randomTestSeed)`.

File: test/tools/fuzzTesting/fuzzHelper.h (hunk 15 above)
Comment (chfast) [reply test/tools/fuzzTesting/fuzzHelper.h]: So also remove this.

File: test/tools/libtesteth/Options.cpp (hunk 19 above)
Comment (winsvega) [reply test/tools/libtesteth/Options.cpp]: m_vrs in transaction signature also using is_initialized. 
I just want it to be clear that it is not the int value being checked but boost::optional

File: test/tools/libtesteth/Options.cpp (hunk 18 above)
Comment (pirapira) [reply test/tools/libtesteth/Options.cpp]: What is hidden? Invalid value errors about negative `int`s?

File: test/tools/fuzzTesting/fuzzHelper.h [hunk 20]
```diff
@@ -44,20 +44,17 @@ struct RlpDebug
...
 	RandomCode(RandomCode const&) = delete;
 	void operator=(RandomCode const&) = delete;
 
 private:
-	RandomCode()
-	{
-		generator.reset(new BoostRandomCode());
-	}
-	std::unique_ptr<RandomCodeBase> generator;
+	RandomCode(){}
```
Comment (chfast) [inline test/tools/fuzzTesting/fuzzHelper.h]: This either should be `RandomCode() = default` or removed.

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 21]
```diff
@@ -243,10 +243,9 @@ bool isOpcodeDefined(uint8_t _opcode)
 {
 	eth::Instruction inst = (eth::Instruction) _opcode;
 	eth::InstructionInfo info = eth::instructionInfo(inst);
-	if (info.name.find("INVALID_INSTRUCTION") != std::string::npos || info.name.empty()
-		|| std::find(invalidOpcodes.begin(), invalidOpcodes.end(), inst) != invalidOpcodes.end())
-		return false;
-	return true;
+	bool isUndefined = (info.name.find("INVALID_INSTRUCTION") != std::string::npos || info.name.empty()
```
Comment (chfast) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: Why not `isDefined = ...; return isDefined`? Double negation is confusing.

File: test/tools/libtesteth/Options.cpp [hunk 22]
```diff
@@ -209,7 +209,10 @@ Options::Options(int argc, char** argv)
 		else if (arg == "--seed")
 		{
 			throwIfNoArgumentFollows();
-			randomTestSeed = static_cast<uint64_t>(min<u256>(UINT64_MAX, toInt(argv[++i])));
+			u256 input = toInt(argv[++i]);
+			if (input > UINT64_MAX)
```
Comment (chfast) [inline test/tools/libtesteth/Options.cpp]: I would stick to `std::numeric_limits<uint64_t>` although longer.
